### ISSUE TITLE
[2.1.0.Beta2] [ES 1.3.2] [Spark 1.1.0] EsHadoopNoNodesLeftException 

### ISSUE THREAD

**Issue by SHSE** [elasticsearch/elasticsearch-hadoop#303]
Full stack trace:
```
org.elasticsearch.hadoop.rest.EsHadoopNoNodesLeftException: Connection error (check network and/or proxy settings)- all nodes failed
 org.elasticsearch.hadoop.rest.NetworkClient.execute(NetworkClient.java:123)
        org.elasticsearch.hadoop.rest.RestClient.execute(RestClient.java:303)
        org.elasticsearch.hadoop.rest.RestClient.execute(RestClient.java:299)
        org.elasticsearch.hadoop.rest.RestClient.scan(RestClient.java:329)
        org.elasticsearch.hadoop.rest.RestRepository.scan(RestRepository.java:123)
        org.elasticsearch.hadoop.rest.QueryBuilder.build(QueryBuilder.java:209)
        org.elasticsearch.hadoop.rest.RestService$PartitionReader.scrollQuery(RestService.java:99)
        org.elasticsearch.spark.rdd.AbstractEsRDDIterator.reader$lzycompute(AbstractEsRDDIterator.scala:35)
        org.elasticsearch.spark.rdd.AbstractEsRDDIterator.reader(AbstractEsRDDIterator.scala:27)
        org.elasticsearch.spark.rdd.AbstractEsRDDIterator.close(AbstractEsRDDIterator.scala:66)
        org.elasticsearch.spark.rdd.AbstractEsRDDIterator.closeIfNeeded(AbstractEsRDDIterator.scala:59)
        org.elasticsearch.spark.rdd.AbstractEsRDDIterator$$anonfun$1.apply$mcV$sp(AbstractEsRDDIterator.scala:39)
        org.elasticsearch.spark.rdd.AbstractEsRDDIterator$$anonfun$1.apply(AbstractEsRDDIterator.scala:39)
        org.elasticsearch.spark.rdd.AbstractEsRDDIterator$$anonfun$1.apply(AbstractEsRDDIterator.scala:39)
        org.elasticsearch.spark.rdd.CompatUtils$Spark11TaskContext$1.onTaskCompletion(CompatUtils.java:47)
        org.apache.spark.TaskContext$$anonfun$markTaskCompleted$1.apply(TaskContext.scala:107)
        org.apache.spark.TaskContext$$anonfun$markTaskCompleted$1.apply(TaskContext.scala:107)
        scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
        org.apache.spark.TaskContext.markTaskCompleted(TaskContext.scala:107)
        org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:77)
        org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
        org.apache.spark.scheduler.Task.run(Task.scala:54)
        org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:177)
        java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        java.lang.Thread.run(Thread.java:745)
```

**Comment by costin** [comment]
@b0c1 I've ran the test suite with queries such as `?q=name:xxx` or `?q=foo:ba*` and everything works out fine.
What is your configuration - both queries and nodes? The exception above indicates the nodes are incorrect (`null`) which might be or not related to the query itself. Does this occur with Spark or M/R ?

**Comment by b0c1** [comment]
@costin Nothing extra. If I set es.query parameter to the esRDD, I got the error (if not set it's work fine). As you see I'm using spark :)
My code is simple:
```scala
    val params = Map(
      ConfigurationOptions.ES_NODES -> config.esHost,
      ConfigurationOptions.ES_PORT -> config.esHttPort.toString,
      ConfigurationOptions.ES_MAPPING_ID -> "id",
      ConfigurationOptions.ES_BATCH_SIZE_BYTES -> "10mb",
      ConfigurationOptions.ES_QUERY -> query,
      ConfigurationOptions.ES_INPUT_JSON -> "true",
      ConfigurationOptions.ES_RESOURCE_READ -> s"*/enriched"
    )

    val xx = EsSpark.esRDD(sc, params)
```

**Comment by costin** [comment]
@b0c1 What are the values of `config.esHost`, `config.esHttPort.toString` (typo? - Http vs Htt) and `query`?

**Comment by b0c1** [comment]
@costin I created a local es server , so the host is "localhost", the http port is random binded port. I using this code to pull up es server:

```scala
object TestElasticNode extends Logging {

  val tempFile = File.createTempFile("elasticsearchtests", "tmp")
  val homeDir = new File(tempFile.getParent + "/" + UUID.randomUUID().toString)
  homeDir.mkdir()
  homeDir.deleteOnExit()
  tempFile.deleteOnExit()
  logger.info("Setting ES home dir [{}]", homeDir)

  private def getFreePort(): Int = {
    val ss = new ServerSocket(0)
    val port = ss.getLocalPort
    ss.close()
    port
  }

  val esHttpPort = getFreePort()
  val esTransportPort = getFreePort()

  val settings = ImmutableSettings.settingsBuilder()
    .put("network.host", "127.0.0.1")
    .put("node.http.enabled", true)
    .put("http.enabled", true)
    .put("http.port", esHttpPort)
    .put("transport.tcp.port", esTransportPort)
    .put("transport.publish_port", esTransportPort)
    .put("path.home", homeDir.getAbsolutePath)
    .put("index.number_of_shards", 1)
    .put("index.number_of_replicas", 0)
    .put("script.disable_dynamic", false)
    .put("es.logger.level", "DEBUG")

  implicit val client = ElasticClient.fromNode(NodeBuilder.nodeBuilder().local(false).data(true).settings(settings).node())
  client.client.admin().cluster().prepareHealth().setWaitForGreenStatus().execute().actionGet()
}
```

if the `query` is `empty string` => the code work, no error
if the `query` is `s"?q=timestamp:>=${from}"` => I got an error (from is a long value calculated by spark startup parameter (example fetch the last one hour data))

**Comment by costin** [comment]
@b0c1 I'm afraid I can't reproduce the error despite my numerous attempts. See the latest test added to the suite [here](https://github.com/elasticsearch/elasticsearch-hadoop/blob/027c127044a80521ee2d6cc06919ccbb7f428346/spark/src/itest/scala/org/elasticsearch/spark/integration/AbstractScalaEsSpark.scala#L161). With or without the colon `:`, the query works just fine.
Based on the logs
```
Invalid target URI POST@null/*/enriched/_search?q=timestamp:>=1418769000&search_type=scan&scroll=5&size=50&preference=_shards:0;_only_node:nJg1MFJbQPiZL8BSMbLmIw] failed (localhost:63053); no other nodes left - aborting...
```
it looks your host/port is passed as `null` (instead of `localhost:9200` or what have you). Or potentially it becomes invalid due to the use of `:` though I can't reproduce this ...

**Comment by b0c1** [comment]
@costin Hi bro, I don't think that... maybe something weird ... look the log message, it's contain:
`failed (localhost:63053)` -> so it's know the port... maybe try to set your port and es to other than default. maybe the asterisk in the resource? (I set only the ES_RESOURCE_READ to */enriched not the ES_RESOURCE)

**Comment by costin** [comment]
I've tried with or without `*` with `127.0.0.1` and `localhost`, without a port and host and with some. Can you please try the latest binary (mentioned that in #283) and report back. If the error persist please turn on logging on `org.elasticsearch.hadoop` all the way to `TRACE` and ping me back.
Thanks!

**Comment by b0c1** [comment]
@costin if you release beta4 to repository... or if you give me a snapshot repo address ;)

**Comment by costin** [comment]
The dev builds have been published already (sorry, I've mentioned this in #283) so you can download them from Maven. Try the usual `2.1.0.BUILD-SNAPSHOT` as explained [here](http://www.elasticsearch.org/guide/en/elasticsearch/hadoop/master/install.html#download-dev)

**Comment by b0c1** [comment]
@costin I checked, and it's work with SNAPSHOT.

**Comment by costin** [comment]
great - probably that's why I couldn't reproduce it. Can you please check out the fix for #283 as well? Thanks!

**Comment by costin** [comment]
Closing the issue...

**Comment by deepakas** [comment]
Hi Costin, I am getting the same error. 
 WARN Utils: Service 'SparkUI' could not bind on port 4141. Attempting port 4142.
15/07/05 16:49:34 ERROR NetworkClient: Node [Invalid target URI GET@null/"twitter/datascience"/_mapping] failed (127.0.0.1:9200); no other nodes left - aborting...
Exception in thread "main" org.elasticsearch.hadoop.rest.EsHadoopNoNodesLeftException: Connection error (check network and/or proxy settings)- all nodes failed; tried [[127.0.0.1:9200]] 

It works with using val esRDD = sc.esJsonRDD("test/datascience")

I am getting error for this. 
val options13 = Map("path" -> "test/datascience",
      "pushdown" -> "true",
      "es.nodes" -> "127.0.0.1", "es.port" -> "9200",
      "es.index.auto.create" -> "true")

    // Spark 1.3 style
    val df = sql.load("org.elasticsearch.spark.sql", options13)

Also getting a different error while using DF
    val df = sql.esDF("test/datascience",options13)

**Comment by nollbit** [comment]
Me too.

```
Invalid target URI POST@null/zipkin-2016-10-24/span/_search?search_type=scan&scroll=5m&size=50&preference=_shards:2|_local````

That query does not pass the URI test, because of the `|` (pipe) [1] [2]:
```        try {
            // validate new URI
            uri = http.getURI().toString();
        } catch (URIException uriex) {
            throw new EsHadoopTransportException("Invalid target URI " + request, uriex);
        }
```

[1] from https://github.com/elastic/elasticsearch-hadoop/blob/8e44fc65d15d27f29379f8e9a7db67b43d301dd7/mr/src/main/java/org/elasticsearch/hadoop/rest/SearchRequestBuilder.java#L197

[2] as verified by 
```
    val h = new PostMethod
    h.setPath("/zipkin-2016-10-24/span/_search?search_type=scan&scroll=5m&size=50&preference=_shards:2|_local")
    h.getURI.toString // -> org.apache.commons.httpclient.URIException: Invalid query
```

**Comment by jbaiera** [comment]
@nollbit this is a known issue and was addressed in #874, #877, and #868.
